micronaut-jms-core/src/main/java/io/micronaut/jms/pool/AbstractPool.java

```diff
@@ -0,0 +1,69 @@
+package io.micronaut.jms.pool;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+public abstract class AbstractPool<T extends PooledObject<?>> {
```

Comment:
@graemerocher unfortunately plain Apache Commons Pool won't work since we need to pass arguments when we create Sessions and Producers. Thankfully there is an [existing project](https://github.com/fusesource/jmspool) however unfortunately that has been merged into [ActiveMQ](https://github.com/apache/activemq/tree/master/activemq-jms-pool/src/main/java/org/apache/activemq/jms/pool) so we'll have to basically re-implement the same things in this project. I'll write it in here for now but might pull it out into a separate library in the future